### PR TITLE
update bpe tokenizer, lower-case special tokens

### PR DESCRIPTION
**Pull request by alexholdenmiller**
- switch bpe helper to use final dictionary counts instead of remembering all training data
- replace --bpe-num-symbols with --dict-maxtokens
- use --dict-minfreq for bpe's min frequency
- switch special tokens to lower case, as upper case causes frequent issues with lookups (if using --dict-lower(), `dict['__START__']` returns the index of the start token but `dict.txt2vec('__START__')` returns the unknown index because it doesn't recognize `'__start__'`)
- convert loaded upper-cased special tokens to lower case

to handle the bug described we could put a `lower()` in the indexing function, but the place it is now makes reasonable sense (at the beginning of tokenizing)

### CODE REVIEW

File: parlai/core/dict.py [hunk 1]
```diff
@@ -383,46 +377,45 @@ def tokenize(self, text, building=False):
...
-                # other dicts can be modified as we go
-                idx = self.tok2ind.pop(token)
-                del self.ind2tok[idx]
+
         for token in to_remove:
             del self.freq[token]
+            idx = self.tok2ind.pop(token)
+            del self.ind2tok[idx]
+
+    def remove_non_bpe(self):
```
Comment (stephenroller) [inline parlai/core/dict.py]: should this be `_remove_non_bpe`? It seems like a good idea to signal to users that this isn't something they should call themselves.

File: parlai/core/dict.py [hunk 2]
```diff
@@ -579,82 +586,69 @@ class _BPEHelper(object):
...
-    def _apply(self, tokens):
-        return self.bpe.segment_tokens(tokens)
+        if hasattr(self, 'bpe'):
+            return self.bpe.segment_tokens(tokens)
+        else:
+            return tokens
 
-    def finalize(self):
-        """Build the codecs"""
-        if self.built:
```
Comment (stephenroller) [inline parlai/core/dict.py]: I'm worried about this, b/c without testing this code, it looks to me like loading a model from disk will still cause the BPE encoder to retrain. This is a very expensive step, and so if the dict and codecs files are already built, we should use them.

File: parlai/core/dict.py [hunk 3]
```diff
@@ -116,7 +116,7 @@ def add_cmdline_args(argparser):
             dictionary.add_argument(
                 '--dict-maxtokens', default=DictionaryAgent.default_maxtokens,
                 type=int,
-                help='max number of tokens to include in sorted dict')
+                help='max number of tokens to include in dictionary or bpe codecs.')
```
Comment (stephenroller) [inline parlai/core/dict.py]: should we make a similar note for minfreq?

File: parlai/core/dict.py [hunk 4]
```diff
@@ -460,8 +459,15 @@ def save(self, filename=None, append=False, sort=True):
         """
         filename = self.opt['dict_file'] if filename is None else filename
         print('Dictionary: saving dictionary to {}'.format(filename))
-        if sort:
-            self.sort()
+
+        if self.tokenizer == 'bpe':
+            self.bpehelper.finalize((f'{k} {v}' for k, v in self.freq.items()),
```
[the remark this answers is not in the document]
Comment (alexholdenmiller) [reply parlai/core/dict.py]: good call!

File: parlai/core/dict.py (hunk 1 above)
Comment (alexholdenmiller) [reply parlai/core/dict.py]: agreed, good call

File: parlai/core/dict.py (hunk 3 above)
Comment (alexholdenmiller) [reply parlai/core/dict.py]: done

**Pull request merged**